src/mono/mono/component/debugger-agent.c

```diff
@@ -5416,11 +5432,184 @@ static ErrorCode decode_fixed_size_array_internal (MonoType *t, int type, MonoDo
+		ret += sizeof(gunichar2);
+		break;
+	case MONO_TYPE_I2:
+	case MONO_TYPE_U2:
+		decode_int (buf, &buf, limit);
+		ret += sizeof(guint16);
+		break;
+	case MONO_TYPE_I4:
+	case MONO_TYPE_U4:
+	case MONO_TYPE_R4:	
+		decode_int (buf, &buf, limit);
+		ret += sizeof(guint32);
+		break;
+	case MONO_TYPE_I8:
+	case MONO_TYPE_U8:	
+	case MONO_TYPE_R8:
+		decode_long (buf, &buf, limit);
+		ret += sizeof(guint64);
+		break;
+	case MONO_TYPE_PTR:
+	case MONO_TYPE_FNPTR:
+		decode_long (buf, &buf, limit);
+		ret += sizeof(gssize);
+		break;
+	case MONO_TYPE_GENERICINST:
+		if (MONO_TYPE_ISSTRUCT (t)) {
+			/* The client sends these as a valuetype */
+			goto handle_vtype;
+		} else {
+			goto handle_ref;
+		}
+		break;
+	case MONO_TYPE_I:
+	case MONO_TYPE_U:
+		/* We send these as vtypes, so we get them back as such */
+		g_assert (type == MONO_TYPE_VALUETYPE);
+		/* Fall through */
+		handle_vtype:
+	case MONO_TYPE_VALUETYPE:
+		if (type == MONO_TYPE_OBJECT || type == MONO_TYPE_STRING) {
+			/* Boxed vtype */
+			decode_objid (buf, &buf, limit);
+		} else {
+			ret += decode_vtype_compute_size (t, domain, buf, &buf, limit, from_by_ref_value_type);
+		}
+		break;
+	handle_ref:
+	default:
+		if (MONO_TYPE_IS_REFERENCE (t)) {
+			if (type == MONO_TYPE_CLASS || type == MONO_TYPE_OBJECT || type == MONO_TYPE_STRING) {
+				ret += sizeof(MonoObject*);
+				decode_objid (buf, &buf, limit);
+			} else if (type == VALUE_TYPE_ID_NULL) {
+				ret += sizeof(MonoObject*);
+				if (CHECK_PROTOCOL_VERSION (2, 59)) {
+					decode_byte (buf, &buf, limit);
+					decode_int (buf, &buf, limit); //not used
+				}
+			} else if (type == MONO_TYPE_VALUETYPE) {
+				MonoClass *klass;
+				MonoDomain *d;
+				decode_byte (buf, &buf, limit);
+				if (CHECK_PROTOCOL_VERSION(2, 61))
+					decode_byte (buf, &buf, limit); //ignore is boxed
+				klass = decode_typeid (buf, &buf, limit, &d, &err);
+				ret += decode_vtype_compute_size (NULL, domain, buf, &buf, limit, from_by_ref_value_type);
+			} else {
+				goto end;
+			}
+		} else if ((t->type == MONO_TYPE_GENERICINST) &&
+					mono_metadata_generic_class_is_valuetype (t->data.generic_class) &&
+					m_class_is_enumtype (t->data.generic_class->container_class)){
+			ret += decode_vtype_compute_size (t, domain, buf, &buf, limit, from_by_ref_value_type);
+		} else {
+			NOT_IMPLEMENTED;
+		}
+		break;
+	}
+end:
+	*endbuf = buf;
+	return ret;
+}
+
 static ErrorCode
-decode_value_internal (MonoType *t, int type, MonoDomain *domain, guint8 *addr, guint8 *buf, guint8 **endbuf, guint8 *limit, gboolean check_field_datatype)
+decode_value_internal (MonoType *t, int type, MonoDomain *domain, guint8 *addr, guint8 *buf, guint8 **endbuf, guint8 *limit, gboolean check_field_datatype, guint8 **extra_space, gboolean from_by_ref_value_type)
 {
 	ErrorCode err;
 
+	if (m_type_is_byref (t)) {
+		g_assert (extra_space != NULL && *extra_space != NULL);
+		*(guint8**)addr = *extra_space;
+		guint8 *buf_int = buf;
+		addr = *(guint8**)addr;
```

Comment:
This is confusing. (Maybe a comment is enough).

This is what I think it's doing:
1. We're going to decode a value
2. The value will be placed into the extra space we allocated starting at `*extra_space`.
3. At the beginning, `addr` is a pointer to some storage. We will write into that storage the address of the value we're decoding.
4. Then we update `addr` to also point at the beginning of the extra space, because this is where the fields of the value will be stored.

Finally, we bump `*extra_space` so that if we need additional extra storage, it will use the extra space after the current value.